### PR TITLE
SOLR-15653: Fix collection creation race that assumes a local cluster…

### PR DESCRIPTION
**Pull request by markrmiller**
…state when the collection has only just been created.

https://issues.apache.org/jira/browse/SOLR-15653

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java [hunk 1]
```diff
@@ -180,7 +180,7 @@ private static int defaultCounterValue(DocCollection collection, boolean newColl
     if (newCollection) return 0;
 
     int defaultValue;
-    if (collection.getSlice(shard) != null && collection.getSlice(shard).getReplicas().isEmpty()) {
+    if (collection == null || collection.getSlice(shard) != null && collection.getSlice(shard).getReplicas().isEmpty()) {
```
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:183]:
```suggestion
    if (collection == null || (collection.getSlice(shard) != null && collection.getSlice(shard).getReplicas().isEmpty())) {
```

or doing `if (newCollection || collection == null) return 0;` at line 180 above might be slightly easier to comprehend.

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java [hunk 2]
```diff
@@ -201,20 +201,20 @@ private static int defaultCounterValue(DocCollection collection, boolean newColl
...
+    int replicaNum = incAndGetId(stateManager, collectionName, defaultValue);
+    String coreName = buildSolrCoreName(collectionName, shard, type, replicaNum);
+    while (collection != null && existCoreName(coreName, collection.getSlice(shard))) {
+      replicaNum = incAndGetId(stateManager, collectionName, defaultValue);
+      coreName = buildSolrCoreName(collectionName, shard, type, replicaNum);
     }
     return coreName;
   }
 
   public static String buildSolrCoreName(DistribStateManager stateManager, DocCollection collection, String shard, Replica.Type type) {
```
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:216]: Whilst following the caller chain here I noticed that https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.10.0/solr/core/src/java/org/apache/solr/cloud/api/collections/MoveReplicaCmd.java#L95-L97 (and other places) needlessly check null-ness after the `ClusterState.getCollection(collection)` call. Out of scope for this pull request that is though.

**Review by cpoerschke — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java (hunk 1 above)
Comment (markrmiller) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:183]: 👍

Comment (cpoerschke) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:183]:
```suggestion
    if (collection.getSlice(shard) != null && collection.getSlice(shard).getReplicas().isEmpty()) {
```

now that line 180 has the null check.

**Pull request closed without merging**